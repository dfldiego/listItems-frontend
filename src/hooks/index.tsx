export { default as useForm } from "./useForm";
export { default as useIsMounted } from "./useIsMounted";
export { default as useQueryString } from "./useQueryString";
export { default as useItem } from "./useItem";
export { default as useItems } from "./useItems";
