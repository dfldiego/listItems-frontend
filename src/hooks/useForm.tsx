import { useState } from "react";

const useForm = <T extends Object>(initialForm: T) => {
  const [formState, setFormState] = useState(initialForm);

  const onInputChange = (name: string, value: string) => {
    setFormState({
      ...formState,
      [name]: value,
    });
  };

  const onResetForm = () => {
    setFormState(initialForm);
  };

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm,
  };
};

export default useForm;
