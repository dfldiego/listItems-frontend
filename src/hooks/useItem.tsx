import { useCallback, useEffect, useState } from "react";

import { getItemByIdAPI } from "../api/items";
import { defaultError } from "../constants";
import { ProductDetail } from "../interfaces";
import useIsMounted from "./useIsMounted";

const useItem = (id: string | undefined) => {
  const isMounted = useIsMounted();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [itemResponse, setItemResponse] = useState<ProductDetail | null>(null);

  const getItem = useCallback(async () => {
    if (!id) return;
    try {
      setLoading(true);
      const response = await getItemByIdAPI(id as string);
      if (isMounted() && response) {
        setItemResponse(response.data.response);
        setError("");
        setLoading(false);
      }
    } catch (error: any) {
      setError(error.response.data.msg ?? defaultError);
      setLoading(false);
    }
  }, [id, isMounted]);

  useEffect(() => {
    getItem();
  }, [getItem]);

  return {
    itemResponse,
    error,
    loading,
  };
};

export default useItem;
