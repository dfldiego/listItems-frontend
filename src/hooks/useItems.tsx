import { useContext } from "react";
import { ItemsContext } from "../context/itemsContext";

const useItems = () => useContext(ItemsContext);

export default useItems;
