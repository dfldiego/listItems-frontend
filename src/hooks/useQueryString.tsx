import { useLocation } from "react-router-dom";
import queryString from "query-string";

const useQueryString = () => {
  const location = useLocation();
  const query = queryString.parse(location.search);
  return {
    ...query,
  };
};

export default useQueryString;
