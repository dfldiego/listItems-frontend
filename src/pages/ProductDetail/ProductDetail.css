@media (min-width: 1000px) {
  .productdetail_firstColumn {
    max-width: 68rem;
    max-height: 68rem;
    grid-column: 1 / 7;
    grid-row: 1 / 6;
    padding-left: 3.2rem;
  }
  .productdetail_secondColumn {
    grid-column: 8 / 11;
    grid-row: 1 / 6;
  }
  .productdetail_second_row {
    grid-column: 1 / 7;
    grid-row: 8 / 11;
    padding-left: 3.2rem;
  }
  .cardItem_img_detail {
    width: 68rem;
    height: 68rem;
  }
}
.cardItem_img_detail {
  width: 100%;
  height: 100%;
}
.productdetail_condition_sold_quantity {
  padding-top: 3.2rem;
  font-size: 1.4rem;
}
.productdetail_title {
  margin-top: 1.6rem;
  font-weight: bold;
  font-size: 2.4rem;
}
.productdetail_amount {
  padding-top: 3.2rem;
  font-size: 4.6rem;
}
.productdetail_amount_decimals {
  font-size: 2.6rem;
  position: relative;
  top: -14px;
}
/* BUTTON */
.productdetail_btn {
  padding-top: 3.2rem;
  padding-right: 3.2rem;
}
.button_buy {
  background: var(--blue);
  border: 1px solid var(--blue);
  border-radius: 0.5rem;
  color: var(--white);
  font-size: 1.6rem;
  padding: 1.2rem 10rem;
  cursor: pointer;
}
/* DESCRIPTION */
.productdetail_description_title {
  font-size: 2.8rem;
  padding-bottom: 3.2rem;
}
.productdetail_description {
  font-size: 1.6rem;
  padding-bottom: 3.2rem;
  color: var(--grey-moderate);
}
