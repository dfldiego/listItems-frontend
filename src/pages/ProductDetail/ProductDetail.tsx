import { useParams, useNavigate } from "react-router-dom";

import "./ProductDetail.css";
import { useItem } from "../../hooks";
import Search from "../../components/search/Search";
import Breadcrumbs from "../../components/breadcrumbs/Breadcrumbs";
import Error from "../../components/error/Error";
import Loading from "../../components/loading/Loading";

const ProductDetail = () => {
  const params = useParams();
  const navigate = useNavigate();
  const { itemResponse, error, loading } = useItem(params.id);

  const handleClick = () => {
    navigate(-1);
  };

  return (
    <div className="page_container">
      <div className="page_search">
        <Search />
      </div>
      {!error && !loading && (
        <>
          <div className="page_breadcrumbs">
            <Breadcrumbs />
          </div>
          {itemResponse !== null &&
            Object.keys(itemResponse.item).length > 0 && (
              <div className="page_content">
                <div className="productdetail_firstColumn">
                  <img
                    src={itemResponse.item.picture}
                    alt={`${itemResponse.item.title}`}
                    className="cardItem_img_detail"
                    data-testid={"item-detail-img-test-id"}
                  />
                </div>
                <div className="productdetail_secondColumn">
                  <div className="productdetail_condition_sold_quantity">
                    {itemResponse.item.condition === "new" && (
                      <span>
                        {`Nuevo - ${itemResponse.item.sold_quantity} vendidos`}
                      </span>
                    )}
                  </div>
                  <div className="productdetail_title">
                    <span>{itemResponse.item.title}</span>
                  </div>
                  <div className="productdetail_amount">
                    <span>$ {itemResponse.item.price.amount}</span>
                    <span className="productdetail_amount_decimals">
                      {(itemResponse.item.price.decimals % 1)
                        .toFixed(2)
                        .slice(2)}
                    </span>
                  </div>
                  <div className="productdetail_btn">
                    <button
                      className="button_buy"
                      data-testid="button-test-id"
                      onClick={handleClick}
                    >
                      Confirmar
                    </button>
                  </div>
                </div>
                <div className="productdetail_second_row">
                  <div className="productdetail_description_title">
                    <span>Descripcion del producto</span>
                  </div>
                  <div className="productdetail_description">
                    <span>{itemResponse.item.description}</span>
                  </div>
                </div>
              </div>
            )}
        </>
      )}
      <Error loading={loading} error={error} />
      <Loading loading={loading} />
    </div>
  );
};

export default ProductDetail;
