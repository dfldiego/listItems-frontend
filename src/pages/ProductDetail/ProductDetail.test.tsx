import { render, screen, fireEvent } from "@testing-library/react";
import axios from "axios";
import { useItem } from "../../hooks";
import * as router from "react-router-dom";
import ProductDetail from "./ProductDetail";

// useQueryString
jest.mock("../../hooks/useQueryString", () => {
  return () => ({
    q: "remera",
  });
});

// axios
jest.mock("axios");
// @ts-ignore
axios.get.mockResolvedValue({ data: { id: "123", name: "Example Item" } });

// useNavigate
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

// useItems (mock de custom hook con useContext)
jest.mock("../../hooks/useItems.tsx", () => {
  return () => ({
    itemsState: {
      breadcrumbs: ["comida", "postre"],
    },
  });
});

// useItem (mock custom hook)
jest.mock("../../hooks");
const mockUseItem = useItem as jest.MockedFunction<typeof useItem>;
const mockUseItemReturn = {
  itemResponse: {
    author: {
      nickname: null,
    },
    item: {
      id: "id_data",
      title: "title_data",
      price: {
        currency: "currency_data",
        amount: 110,
        decimals: 0.5,
      },
      picture: "picture_data",
      condition: "new",
      free_shipping: true,
      sold_quantity: 500,
      description: "description_data",
    },
  },
  error: "",
  loading: false,
};

describe("<ProductDetail>", () => {
  const navigate = jest.fn();
  beforeEach(() => {
    jest.spyOn(router, "useNavigate").mockImplementation(() => navigate);
  });
  test("should render correctly", () => {
    mockUseItem.mockReturnValue(mockUseItemReturn);
    const { container } = render(<ProductDetail />);
    expect(container).toMatchSnapshot();
  });

  test("should contain search component", () => {
    mockUseItem.mockReturnValue(mockUseItemReturn);
    render(<ProductDetail />);
    const searchComponent = screen.getByTestId("search-component-test-id");
    expect(searchComponent).toBeInTheDocument();
  });

  test("should render breadcrumbs and item detail data", () => {
    mockUseItem.mockReturnValue(mockUseItemReturn);
    render(<ProductDetail />);
    const breadcrumbsComponent = screen.getByTestId(
      "breadcrumbs-component-test-id"
    );
    expect(breadcrumbsComponent).toBeInTheDocument();
    //title
    expect(screen.getByText(/title_data/i)).toBeInTheDocument();
    //amount
    expect(screen.getByText(/110/i)).toBeInTheDocument();
    //decimal
    expect(screen.getByText("50")).toBeInTheDocument();
    //img
    const img = screen.getByTestId("item-detail-img-test-id");
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute("src", "picture_data");
    //description
    expect(screen.getByText(/description_data/i)).toBeInTheDocument();
    //sold quantity
    expect(screen.getByText(/Nuevo - 500 vendidos/i)).toBeInTheDocument();
  });

  test("should contain a button", () => {
    mockUseItem.mockReturnValue(mockUseItemReturn);
    render(<ProductDetail />);
    const buttonElement = screen.getByTestId("button-test-id");
    expect(buttonElement).toBeInTheDocument();
    fireEvent.click(buttonElement);
    expect(navigate).toHaveBeenCalled();
  });
});
