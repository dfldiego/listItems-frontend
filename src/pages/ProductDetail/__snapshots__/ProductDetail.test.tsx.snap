// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetail> should render correctly 1`] = `
<div>
  <div
    class="page_container"
  >
    <div
      class="page_search"
    >
      <div
        class="search_container"
        data-testid="search-component-test-id"
      >
        <div
          class="search_logo"
        >
          <img
            alt="MercadoLibre_logo"
            class="search_logo_img"
            data-testid="meli-logo-test-id"
            src="Logo_ML.png"
          />
        </div>
        <div
          class="search_form"
        >
          <form
            aria-label="form"
            class="search_form_row"
          >
            <input
              autocomplete="off"
              class="input-text"
              name="searchText"
              placeholder="Nunca dejes de buscar"
              type="text"
              value="remera"
            />
            <button
              class="search_button"
              type="submit"
            >
              <div>
                <img
                  alt="Logotipo Buen Sabor"
                  class="search_img"
                  src="ic_Search.png"
                />
              </div>
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="page_breadcrumbs"
    >
      <ul
        class="breadcrumbs_ul"
        data-testid="breadcrumbs-component-test-id"
      >
        <li
          class="breadcrumbs_li"
        >
          <span
            class="breadcrumbs_item"
          >
            comida
          </span>
        </li>
        <li
          class="breadcrumbs_li"
        >
          <span
            class="breadcrumbs_item_bold"
          >
            postre
          </span>
        </li>
      </ul>
    </div>
    <div
      class="page_content"
    >
      <div
        class="productdetail_firstColumn"
      >
        <img
          alt="title_data"
          class="cardItem_img_detail"
          data-testid="item-detail-img-test-id"
          src="picture_data"
        />
      </div>
      <div
        class="productdetail_secondColumn"
      >
        <div
          class="productdetail_condition_sold_quantity"
        >
          <span>
            Nuevo - 500 vendidos
          </span>
        </div>
        <div
          class="productdetail_title"
        >
          <span>
            title_data
          </span>
        </div>
        <div
          class="productdetail_amount"
        >
          <span>
            $ 
            110
          </span>
          <span
            class="productdetail_amount_decimals"
          >
            50
          </span>
        </div>
        <div
          class="productdetail_btn"
        >
          <button
            class="button_buy"
            data-testid="button-test-id"
          >
            Confirmar
          </button>
        </div>
      </div>
      <div
        class="productdetail_second_row"
      >
        <div
          class="productdetail_description_title"
        >
          <span>
            Descripcion del producto
          </span>
        </div>
        <div
          class="productdetail_description"
        >
          <span>
            description_data
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
